### PR TITLE
Provide a mean to dump and restore the encoder config

### PR DESCRIPTION
**Pull request by lu-zero**
It is useful to debug so maybe we could use it in another hidden subcommand, right now I put it in the `advanced` one.

``` sh
# cargo run -- advanced --help
rav1e-advanced
Advanced features

USAGE:
    rav1e advanced [OPTIONS]

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -l, --load_config <LOAD_CONFIG>    Save the encoder configuration from a toml file
    -s, --save_config <SAVE_CONFIG>    Save the current configuration in a toml file
    -c, --completion <SHELL>           Output to stdout the completion definition for the shell [possible values: zsh,
                                       bash, fish, powershell, elvish]
```

It should make easier to reproduce specific scenarios and avoid repeating manually what is set or not set in the speed levels.

```
   -s, --speed <SPEED>
            Speed level (0 is best quality, 10 is fastest)
            Speeds 10 and 0 are extremes and are generally not recommended
            -0:
            min_block_size = BLOCK_4X4, multiref = true, fast_deblock = false, reduced_tx_set = false,
            tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = true, rdo_tx_decision = true,
            prediction_modes = ComplexAll, include_near_mvs = true, no_scene_detection = false, diamond_me = true, cdef
            = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = true
            -1:
            min_block_size = BLOCK_4X4, multiref = true, fast_deblock = false, reduced_tx_set = false,
            tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = true, rdo_tx_decision = true,
            prediction_modes = ComplexAll, include_near_mvs = true, no_scene_detection = false, diamond_me = true, cdef
            = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = false
            -2:
            min_block_size = BLOCK_4X4, multiref = true, fast_deblock = false, reduced_tx_set = false,
            tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = false, rdo_tx_decision = true,
            prediction_modes = ComplexAll, include_near_mvs = true, no_scene_detection = false, diamond_me = true, cdef
            = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = false
            -3:
            min_block_size = BLOCK_8X8, multiref = true, fast_deblock = false, reduced_tx_set = false,
            tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = false, rdo_tx_decision = true,
            prediction_modes = ComplexKeyframes, include_near_mvs = true, no_scene_detection = false, diamond_me = true,
            cdef = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = false
            -4:
            min_block_size = BLOCK_8X8, multiref = true, fast_deblock = false, reduced_tx_set = false,
            tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = false, rdo_tx_decision = true,
            prediction_modes = ComplexKeyframes, include_near_mvs = false, no_scene_detection = false, diamond_me =
            true, cdef = true, quantizer_rdo = false, use_satd_subpel = true, non_square_partition = false
            -5:
```

### CODE REVIEW

File: src/bin/common.rs [hunk 1]
```diff
@@ -312,6 +313,18 @@ pub fn parse_cli() -> Result<CliOptions, CliError> {
...
                      .possible_values(&Shell::variants())
                 )
+                .arg(Arg::with_name("SAVE_CONFIG")
+                     .help("Save the current configuration in a toml file")
+                     .short("s")
+                     .long("save_config")
+                     .takes_value(true)
+                )
+                .arg(Arg::with_name("LOAD_CONFIG")
+                     .help("Save the encoder configuration from a toml file")
```
Comment (shssoichiro) [inline src/bin/common.rs]: Should be "load" instead of "save".

**Comment by rzumer** [comment]
For issues like #1671 this does not really help, because the changes are hardcoded in the `Sequence`, not `EncoderConfig`. I guess it also won't work when the configuration is generated on a branch that has modified any field in `EncoderConfig` or its descendants, unless tested on the same branch.

How would this feature help debug better than knowing the original command line and sample?

**Comment by lu-zero** [comment]
> For issues like #1671 this does not really help, because the changes are hardcoded in the Sequence, not EncoderConfig. 

That might or might not be interesting to address it (maybe after it does not causes crashes)

> How would this feature help debug better than knowing the original command line and input sample?

It is more compact that I can see, it is more or less and extension of what @shssoichiro did with the speed settings.

File: src/bin/common.rs (hunk 1 above)
Comment (lu-zero) [reply src/bin/common.rs]: Oh right

**Comment by lu-zero** [comment]
@ycho here is what you asked regarding presenting what each speed level presents.

**Comment by ycho** [comment]
> @ycho here is what you asked regarding presenting what each speed level presents.

All right, thanks for the precut. 
I see the UPPERCASE is used for one option, i.e. blocksize.
And... in 3 years I will ask to use "o|x" replacing true|false or yes|no! :) (kidding....)
Always appreciate the trials of improving config interface!

**Comment by lu-zero** [comment]
> > @ycho here is what you asked regarding presenting what each speed level presents.
> 
> All right, thanks for the precut.
> I see the UPPERCASE is used for one option, i.e. blocksize.

The type itself is like that, the serialization rules are _quite_ verbatim to the code, potentially it could be made lower-case.

**Comment by dwbuiten** [comment]
Does this move `serde` from being a debug-only or cli-only dependency to a hard dependency for the API? If so, a 👎 from me.

**Comment by lu-zero** [comment]
Now it is fully optional, we can discuss making the option always on for the cli.

**Comment by ycho** [comment]
> It should make easier to reproduce specific scenarios and avoid repeating manually what is set or not set in the speed levels.
> 
> ```
>    -s, --speed <SPEED>
>             Speed level (0 is best quality, 10 is fastest)
>             Speeds 10 and 0 are extremes and are generally not recommended
>             -0:
>             min_block_size = BLOCK_4X4, multiref = true, fast_deblock = false, reduced_tx_set = false,
>             tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = true, rdo_tx_decision = true,
>             prediction_modes = ComplexAll, include_near_mvs = true, no_scene_detection = false, diamond_me = true, cdef
>             = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = true
>             -1:
>             min_block_size = BLOCK_4X4, multiref = true, fast_deblock = false, reduced_tx_set = false,
>             tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = true, rdo_tx_decision = true,
>             prediction_modes = ComplexAll, include_near_mvs = true, no_scene_detection = false, diamond_me = true, cdef
>             = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = false
>             -2:
>             min_block_size = BLOCK_4X4, multiref = true, fast_deblock = false, reduced_tx_set = false,
>             tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = false, rdo_tx_decision = true,
>             prediction_modes = ComplexAll, include_near_mvs = true, no_scene_detection = false, diamond_me = true, cdef
>             = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = false
>             -3:
>             min_block_size = BLOCK_8X8, multiref = true, fast_deblock = false, reduced_tx_set = false,
>             tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = false, rdo_tx_decision = true,
>             prediction_modes = ComplexKeyframes, include_near_mvs = true, no_scene_detection = false, diamond_me = true,
>             cdef = true, quantizer_rdo = true, use_satd_subpel = true, non_square_partition = false
>             -4:
>             min_block_size = BLOCK_8X8, multiref = true, fast_deblock = false, reduced_tx_set = false,
>             tx_domain_distortion = true, tx_domain_rate = false, encode_bottomup = false, rdo_tx_decision = true,
>             prediction_modes = ComplexKeyframes, include_near_mvs = false, no_scene_detection = false, diamond_me =
>             true, cdef = true, quantizer_rdo = false, use_satd_subpel = true, non_square_partition = false
>             -5:
> ```

It might be late but are we getting the help message like this? If so and you don't mind, I'd sincerely volunteer myself to convert this help message to better readable table style! :)

**Comment by lu-zero** [comment]
I made it so it outputs like this:
```
    -s, --speed <SPEED>
            Speed level (0 is best quality, 10 is fastest)
            Speeds 10 and 0 are extremes and are generally not recommended
             0 :
                min_block_size = BLOCK_4X4
                multiref = true
                fast_deblock = false
                reduced_tx_set = false
                tx_domain_distortion = true
                tx_domain_rate = false
                encode_bottomup = true
                rdo_tx_decision = true
                prediction_modes = ComplexAll
                include_near_mvs = true
                no_scene_detection = false
                fast_scene_detection = false
                diamond_me = true
                cdef = true
                quantizer_rdo = true
                use_satd_subpel = true
                non_square_partition = true
                enable_segmentation = false
             1 :
                min_block_size = BLOCK_4X4
                multiref = true
                fast_deblock = false
                reduced_tx_set = false
                tx_domain_distortion = true
                tx_domain_rate = false
                encode_bottomup = true
                rdo_tx_decision = true
                prediction_modes = ComplexAll
                include_near_mvs = true
                no_scene_detection = false
                fast_scene_detection = false
                diamond_me = true
                cdef = true
                quantizer_rdo = true
                use_satd_subpel = true
                non_square_partition = false
                enable_segmentation = false
```

**Comment by ycho** [comment]
> I made it so it outputs like this:
> 
> ```
>     -s, --speed <SPEED>
>             Speed level (0 is best quality, 10 is fastest)
>             Speeds 10 and 0 are extremes and are generally not recommended
>              0 :
>                 min_block_size = BLOCK_4X4
>                 multiref = true
>                 fast_deblock = false
>                 reduced_tx_set = false
>                 tx_domain_distortion = true
>                 tx_domain_rate = false
>                 encode_bottomup = true
>                 rdo_tx_decision = true
>                 prediction_modes = ComplexAll
>                 include_near_mvs = true
>                 no_scene_detection = false
>                 fast_scene_detection = false
>                 diamond_me = true
>                 cdef = true
>                 quantizer_rdo = true
>                 use_satd_subpel = true
>                 non_square_partition = true
>                 enable_segmentation = false
>              1 :
>                 min_block_size = BLOCK_4X4
>                 multiref = true
>                 fast_deblock = false
>                 reduced_tx_set = false
>                 tx_domain_distortion = true
>                 tx_domain_rate = false
>                 encode_bottomup = true
>                 rdo_tx_decision = true
>                 prediction_modes = ComplexAll
>                 include_near_mvs = true
>                 no_scene_detection = false
>                 fast_scene_detection = false
>                 diamond_me = true
>                 cdef = true
>                 quantizer_rdo = true
>                 use_satd_subpel = true
>                 non_square_partition = false
>                 enable_segmentation = false
> ```

Thanks for trying this different format. I was considering whether I can put each option as column then each row marks whether each option is true or not , though one line cannot enumerate those many options thus it will need several tables.

**Pull request merged**